### PR TITLE
CBMC proofs for hash_callback functions

### PR DESCRIPTION
**Pull request by danielsn**
*Description of changes:*
CBMC proofs for aws_hash_callback_c_str_eq(), aws_hash_callback_string_destroy(), aws_hash_callback_string_eq(), aws_ptr_eq()

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include <aws/common/string.h>
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+#include <stddef.h>
+
+void aws_hash_callback_c_str_eq_harness() {
+    const char *str1 = make_arbitrary_c_str(MAX_STRING_LEN);
+    const char *str2 = nondet_bool() ? str1 : make_arbitrary_c_str(MAX_STRING_LEN);
+    aws_hash_callback_c_str_eq(str1, str2);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c]: In case it returns 0, could you use `assert_bytes_match`?

File: .cbmc-batch/jobs/aws_hash_callback_c_str_eq/cbmc-batch.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+jobos: ubuntu16
+cbmcflags: "--bounds-check;--div-by-zero-check;--float-overflow-check;--nan-check;--pointer-check;--pointer-overflow-check;--signed-overflow-check;--undefined-shift-check;--unsigned-overflow-check;--unwind;1;--unwinding-assertions;--unwindset;strcmp.0:129;--object-bits;8"
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_hash_callback_c_str_eq/cbmc-batch.yaml]: `--object-bits;8` is default

File: .cbmc-batch/jobs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#include <proof_helpers/make_common_data_structures.h>
+#include <proof_helpers/proof_allocators.h>
+#include <proof_helpers/utils.h>
+#include <stddef.h>
+
+void aws_hash_callback_string_eq_harness() {
+    struct aws_allocator *alloc = can_fail_allocator();
+    const char *str1 = make_arbitrary_aws_string_nondet_len_with_max(alloc, MAX_STRING_LEN);
+    const char *str2 = nondet_bool() ? str1 : make_arbitrary_aws_string_nondet_len_with_max(alloc, MAX_STRING_LEN);
+    aws_hash_callback_string_eq(str1, str2);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c]: In case it returns true, could you use `assert_bytes_match`?

File: .cbmc-batch/jobs/aws_ptr_eq/aws_hash_callback_c_str_eq_harness.c [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/*
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_ptr_eq/aws_hash_callback_c_str_eq_harness.c]: why is this file here?

File: include/aws/common/string.h [hunk 5]
```diff
@@ -220,4 +220,24 @@ AWS_STATIC_IMPL struct aws_byte_cursor aws_byte_cursor_from_string(const struct
     return aws_byte_cursor_from_array(aws_string_bytes(src), src->len);
 }
 
+/**
+ * Best-effort checks data-structure invarients.
+ */
+AWS_STATIC_IMPL bool aws_string_is_valid(const struct aws_string *str) {
+    return str && AWS_MEM_IS_READABLE(&str->bytes[0], str->len + 1) && str->bytes[str->len] == 0;
+    // return str && AWS_MEM_IS_READABLE(str->bytes, str->len + 1) && str->bytes[str->len] == 0;
```
Comment (feliperodri) [inline include/aws/common/string.h]: remove it?

File: .cbmc-batch/jobs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c (hunk 1 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_callback_c_str_eq/aws_hash_callback_c_str_eq_harness.c]: Done.

File: .cbmc-batch/jobs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c (hunk 3 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_callback_string_eq/aws_hash_callback_string_eq_harness.c]: Done

File: include/aws/common/string.h (hunk 5 above)
Comment (danielsn) [reply include/aws/common/string.h]: Done

File: .cbmc-batch/jobs/aws_ptr_eq/aws_hash_callback_c_str_eq_harness.c (hunk 4 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_ptr_eq/aws_hash_callback_c_str_eq_harness.c]: ```cp -r``` error. File removed.

File: .cbmc-batch/jobs/aws_hash_callback_c_str_eq/cbmc-batch.yaml (hunk 2 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_hash_callback_c_str_eq/cbmc-batch.yaml]: Makefile.common autogenerates this file: this should really be a change to Makefile.common, if we care.
